File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
// this adds a constant lookup for each value that will slow us down
// const set = new Set 

// finding an objects properties is faster than searching for a value in a set
  const uniques = {}
  // for each value in arr
  for (const value of arr)
    // if value is repeated
    if (uniques[value]){

      // retur
      return value
    }
    
    // set value to true
    uniques[value] = true
 
  // no duplicates found
  return -1
}

function firstTest(str){
  // array optimized for checking unique values
  const uniques = new Set()
  // each value of array
  for (value of uniques){
    // if value
    if (uniques.has(value)){
    // return
      return value
    }
    // add to array
    uniques.add(value)
  }
  // no duplicates return
  return -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  # make new set
  uniques = Set.new
  # go through array
  arr.each do |value|
    # return if i is included
    return value if uniques.include?(values)
    # or add to uniques
    uniques.add(value)
  end
  # if nothing return no duplicates
  -1
end

def find_first_duplicate(arr)
  # sidebar
# i use a each method that makes an enumarator object giving added flexibility to
# the data such as amount and complexity

  # make set
  uniques = Set.new
  # value added until duplicate returns as nil
  arr.each {|value| return value if uniques.add?(value).nil}
  -1
end
def find_first_duplicate(arr)
  # intialize object
  uniques = {}

  # loop through arr
  for (value of arr){
    # return value if repeated
    return value if uniques[value]
    # put value in object
    uniques[value] = true
  }
  # no duplicates
  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
